### ISSUE TITLE
Improve throughput of db updates

### ISSUE THREAD

**Issue by wolfbeast** [MoonchildProductions/FSyncMS#4]
I've done some research on the structure of InnoDB and how it handles updates, and it seems FSyncMS is using a very expensive "REPLACE INTO" command where it's not really needed.

Apparently, `REPLACE INTO` can be changed to `INSERT INTO ... ON DUPLICATE KEY UPDATE` statements to massively improve the throughput because `REPLACE INTO` doesn't actually update the row; it does a `DELETE` first then an `INSERT`, which requires InnoDB indices to be updated for both operations.

I don't have reference manuals and I'm not comfortable enough with SQL to just tinker with queries. If anyone knows/can confirm how to do this: info and/or PRs are welcome.

**Comment by wolfbeast** [comment]
I think the only command that needs updating for this is the store object one.

https://github.com/MoonchildProductions/FSyncMS/blob/master/weave_storage.php#L223

Someone savvy in SQL please let me know if this looks right?
````
$insert_stmt = 'insert into wbo (username, id, collection, parentid, predecessorid, sortindex, modified, payload, payload_size)
values (:username, :id, :collection, :parentid, :predecessorid, :sortindex, :modified, :payload, :payload_size) 
on duplicate key update 
username=values(username), id=values(id), collection=values(collection), parentid=values(parentid), predecessorid=values(predecessorid), sortindex=values(sortindex), modified=values(modified), payload=value(payload), payload_size=values(payload_size)';
````
And maybe the values staements can be reduced for known-same values (they don't need updating)...?

**Comment by robotman321** [comment]
The syntax is correct. I tested on a temp table for full validation.

**Comment by wolfbeast** [comment]
> The syntax is correct. I tested on a temp table for full validation.

Great! Thanks for confirming the syntax and testing :)

**Comment by wolfbeast** [comment]
Backed out this change because it obviously doesn't do what it is supposed to be doing :( 
With the change nothing gets stored at all in the database in a retrievable form. I must be missing something obvious.

**Comment by wolfbeast** [comment]
Ohboy, talk about a dumb typo. value(payload) instead of values(payload)...

**Comment by wolfbeast** [comment]
Solved by Commit: 540b41055b11b04a861c607cd7c72ce831e7c3bf
Verified working on production, and massively lower IOwait as a result. Considering this closed.

**Comment by stelas** [comment]
Breaks sqlite backend. [Sqlite](http://www.sqlite.org/lang_insert.html) doesn't support "on duplicate key update"... :-1:

**Comment by wolfbeast** [comment]
Patches are welcome to make this conditional!

**Comment by wolfbeast** [comment]
SQLite fixed with a conditional in commit 0615857
